File: src/main/java/com/cartrecipes/controller/CartController.java
package com.cartrecipes.controller;

import com.cartrecipes.controller.dto.cart.CartDto;
import com.cartrecipes.controller.dto.recipe.AddRecipesToCartDto;
import com.cartrecipes.service.cart.CartService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.modelmapper.ModelMapper;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping(CartController.BASE_PATH)
@RequiredArgsConstructor
@Validated
public class CartController {
    public static final String BASE_PATH = "/carts";

    private final CartService cartService;

    private final ModelMapper modelMapper;

    @GetMapping("/{cartId}")
    public CartDto findById(@PathVariable Long cartId) {
        var cart = cartService.findWithRecipesAndProductsById(cartId);
        return modelMapper.map(cart, CartDto.class);
    }

    @PostMapping("/{cartId}/add_recipe")
    public CartDto addRecipes(@PathVariable Long cartId, @RequestBody @Valid AddRecipesToCartDto addRecipesToCartDto) {
        var cart = cartService.addRecipes(cartId, addRecipesToCartDto.recipeIds());
        return modelMapper.map(cart, CartDto.class);
    }

    @DeleteMapping("/{cartId}/recipes/{recipeId}")
    public CartDto removeRecipe(@PathVariable Long cartId, @PathVariable Long recipeId) {
        var cart = cartService.removeRecipe(cartId, recipeId);
        return modelMapper.map(cart, CartDto.class);
    }
}

File: src/main/kotlin/com/cartrecipes/controller/CartController.kt
package com.cartrecipes.controller

import com.cartrecipes.controller.dto.cart.CartDto
import com.cartrecipes.controller.dto.recipe.AddRecipesToCartDto
import com.cartrecipes.service.cart.CartService
import jakarta.validation.Valid
import org.modelmapper.ModelMapper
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping(CartController.BASE_PATH)
@Validated
class CartController (
    private val cartService: CartService,
    private val modelMapper: ModelMapper
) {
    companion object {
        const val BASE_PATH = "/carts"
    }

    @GetMapping("/{cartId}")
    fun findById(@PathVariable cartId: Long): CartDto {
        return cartService.findWithRecipesAndProductsById(cartId)
            .let { modelMapper.map(it, CartDto::class.java) }
    }

    @PostMapping("/{cartId}/add_recipe")
    fun addRecipes(
        @PathVariable cartId: Long,
        @RequestBody @Valid addRecipesToCartDto: AddRecipesToCartDto
    ): CartDto {
        return cartService.addRecipes(cartId, addRecipesToCartDto.recipeIds)
            .let { modelMapper.map(it, CartDto::class.java) }
    }

    @DeleteMapping("/{cartId}/recipes/{recipeId}")
    fun removeRecipe(
        @PathVariable cartId: Long,
        @PathVariable recipeId: Long
    ): CartDto {
        return cartService.removeRecipe(cartId, recipeId)
            .let { modelMapper.map(it, CartDto::class.java) }
    }
}